Support for Durable HTTP Out-of-proc

cgillum opened the pull request:
Additional changes for [Durable HTTP](https://github.com/Azure/azure-functions-durable-extension/issues/469) to make it available to JavaScript orchestrations.

- Enable Durable HTTP out-of-proc
- Fixed some string comparison issues
- Simplified serialization logic

FYI @bachuv, since I changed a lot of your JSON serialization logic, it would be good if you could look over it to make sure I'm not regressing anything.

cgillum commented on src/WebJobs.Extensions.DurableTask/WebJobs.Extensions.DurableTask.csproj:
```diff
@@ -8,7 +8,7 @@
     <MinorVersion>0</MinorVersion>
     <PatchVersion>0</PatchVersion>
-    <Version>$(MajorVersion).$(MinorVersion).$(PatchVersion)-beta1</Version>
+    <Version>$(MajorVersion).$(MinorVersion).$(PatchVersion)-hackathon</Version>
```
Oops. I need to revert this change.

bachuv commented:
@cgillum Looks good to me!

ConnorMcMahon commented on src/WebJobs.Extensions.DurableTask/Listener/TaskOrchestrationShim.cs:
```diff
@@ -148,7 +135,11 @@ public override async Task<string> Execute(OrchestrationContext innerContext, st
                     if (jObj != null)
                     {
-                        await this.HandleOutOfProcExecution(jObj);
+                        bool moreWorkToDo = await this.outOfProcShim.HandleOutOfProcExecutionAsync(jObj);
```
I'm not sure I love splitting out the "more work to do" logic out of `HandleOutOfProcExecutionAsync()`. It seems like that logic was within the scope of that method's responsibility, and it puts more onus on the callers of that method to properly handle the case where there is more work to do.

It looks like this was done for testing purposes. Is there way we can still test this without splitting that logic up?

ConnorMcMahon commented on test/FunctionsV2/OutOfProcTests.cs:
```diff
@@ -0,0 +1,74 @@
+
+            var shim = new OutOfProcOrchestrationShim(contextMock.Object);
+
+            var jsonResponse = @"
```
NIT: this should be `jsonRequest`.

cgillum commented on test/Common/TestFrameworkTests.cs:
```diff
@@ -16,12 +16,13 @@ public class TestFrameworkTests
         {
             Assembly currentAssembly = Assembly.GetExecutingAssembly();
-            List<MethodInfo> testsMissingCategoryTrait = currentAssembly.GetTypes()
+            List<string> testsMissingCategoryTrait = currentAssembly.GetTypes()
```
Yep - this test was failing and I had no idea what was going on. Having a more readable failure message would have saved me some debugging time.

cgillum commented on src/WebJobs.Extensions.DurableTask/DurableHttpResponse.cs:
```diff
@@ -33,47 +30,27 @@ public class DurableHttpResponse
         {
             this.StatusCode = statusCode;
-            this.Headers = headers ?? new Dictionary<string, StringValues>(StringComparer.OrdinalIgnoreCase);
+            this.Headers = HttpHeadersConverter.CreateCopy(headers);
```
My initial reaction is also yes. In this specific case I wanted us to do so because we want to make sure that we're assuming that we have a case-insensitive dictionary, and there's no guarantee that the user would pass us a case-insensitive one. 

Generally speaking, it seems there are two different possibilities of bugs. I assume you're talking about the case where they pass in a dictionary and modify that dictionary in their code afterwards.

* Case 1: The user expects that their dictionary modifications would also modify the state of `DurableHttpResponse`.
* Case 2: The user expects that their dictionary modifications won't affect the state of `DurableHttpResponse` - or they don't think about it at all and just modify the dictionary.

I'm more worried about Case 2 because that's more likely to result in _accidental_ behavior changes. Case 1 seems like a poor/risky assumption on the user's part, so I'm less worried about being defensive about it. If they want to modify the headers afterwards, they should just access the `Headers` property directly.

cgillum commented on src/WebJobs.Extensions.DurableTask/DurableHttpRequest.cs:
```diff
@@ -37,83 +34,81 @@ public class DurableHttpRequest
+        public bool AsynchronousPatternEnabled { get; }
 
-        private static Dictionary<string, StringValues> CreateDictionaryCopy(IDictionary<string, StringValues> headers)
+        private class HttpMethodConverter : JsonConverter
```
`HttpMethod` is actually a class with static properties and not an enum, which is why serialization is a bit trickier.

ConnorMcMahon replied:
Huh, didn't realize that!

cgillum commented on src/WebJobs.Extensions.DurableTask/Listener/OutOfProcOrchestrationShim.cs:
```diff
@@ -0,0 +1,191 @@
+                            ((IDurableOrchestrationContext)this.context).ContinueAsNew(action.Input);
+                            break;
+                        case AsyncActionType.WaitForExternalEvent:
+                            tasks.Add(this.context.WaitForExternalEvent<object>(action.ExternalEventName /*, "ExternalEvent" */));
```
Oh...this might be a hack that I introduced and forgot about. Let me revisit this...

cgillum commented:
I did another iteration of this rebased on top of the latest from the `v2` branch (after merging Sebastian's context refactor PR). I addressed all the PR feedback per my comments so I think this PR should be good to go.

Unfortunately since it's a rebase, you won't be able to see the diffs from previous iterations. :( I wanted to do a rebase since I was merging in other disruptive changes that came in since I started this PR.

BTW, I also opened a follow-up issue that I think we can tackle in v2.0.0-beta3, which is to make the managed identity scenario work for out-of-proc (it's not supported in this PR): https://github.com/Azure/azure-functions-durable-extension/issues/912

cgillum commented on src/WebJobs.Extensions.DurableTask/WebJobs.Extensions.DurableTask.csproj:
```diff
@@ -39,15 +39,15 @@
   <ItemGroup>
     <PackageReference Include="Microsoft.SourceLink.GitHub" Version="1.0.0-*" PrivateAssets="All" />
-    <PackageReference Include="StyleCop.Analyzers" Version="1.1.1-*" PrivateAssets="All" />
+    <PackageReference Include="StyleCop.Analyzers" Version="1.1.118" PrivateAssets="All" />
```
For whatever reason, `1.1.1-*` was not resolving correctly for me. I'll admit I don't really understand this syntax so I'm not sure why it's not happy. Setting it to an explicit version allowed everything to work.

cgillum commented on src/WebJobs.Extensions.DurableTask/Listener/OutOfProcOrchestrationShim.cs:
```diff
@@ -0,0 +1,201 @@
+        /// </summary>
+        /// <param name="executionJson">The result of the out-of-proc execution.</param>
+        /// <returns><c>true</c> if there are more executions to process; <c>false</c> otherwise.</returns>
+        public async Task<bool> ExecuteAsync(JObject executionJson)
```
Just so that it can be called from unit tests. I suppose we could do an `InternalsVisibleTo` thing with the test project, but I usually try to avoid that because it might cause us to take too many dependencies on internal APIs from our test code...
